test/integration/daemonset/daemonset_test.go

```diff
@@ -393,3 +405,221 @@ func TestInsufficientCapacityNodeDaemonDoesNotLaunchPod(t *testing.T) {
+// maxUnavailable requirement is met after a rolling update operation completes
+func checkPodsImageAndAvailability(c clientset.Interface, ds *v1beta1.DaemonSet, image string, maxUnavailable int, podClient corev1typed.PodInterface, t *testing.T) {
+	if err := wait.Poll(pollInterval, pollTimeout, func() (bool, error) {
+		pods := getPods(t, podClient, testLabels())
+		unavailablePods := 0
+		nodesToUpdatedPodCount := make(map[string]int)
+		for _, pod := range pods.Items {
+			if !metav1.IsControlledBy(&pod, ds) {
+				continue
+			}
+			podImage := pod.Spec.Containers[0].Image
+			if podImage == image {
+				nodesToUpdatedPodCount[pod.Spec.NodeName]++
+			}
+			if !podutil.IsPodAvailable(&pod, ds.Spec.MinReadySeconds, metav1.Now()) {
+				unavailablePods++
+			}
+		}
+		if unavailablePods > maxUnavailable {
+			return false, fmt.Errorf("number of unavailable pods %d is greater than maxUnavailable %d", unavailablePods, maxUnavailable)
+		}
+		// Make sure every pod on the node has been updated
+		nodeNames := schedulableNodes(c, ds, t)
+		for _, node := range nodeNames {
+			if nodesToUpdatedPodCount[node] == 0 {
+				return false, nil
+			}
+		}
+		return true, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// A RollingUpdate DaemonSet should rollback without unnecessary restarts
+func TestRollbackWithoutUnnecessaryRestarts(t *testing.T) {
+	server, closeFn, dc, informers, clientset := setup(t)
+	defer closeFn()
+	ns := framework.CreateTestingNamespace("rollback-without-unnecessary-restarts-test", server, t)
+	defer framework.DeleteTestingNamespace(ns, server, t)
+
+	dsClient := clientset.ExtensionsV1beta1().DaemonSets(ns.Name)
+	podClient := clientset.CoreV1().Pods(ns.Name)
+	nodeClient := clientset.CoreV1().Nodes()
+	podInformer := informers.Core().V1().Pods().Informer()
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+	go dc.Run(5, stopCh)
+
+	// Create a rolling update daemonset
+	dsName := "daemonset"
+	ds := newDaemonSet(dsName, ns.Name)
+	ds.Spec.UpdateStrategy = *newRollbackStrategy()
+	oldImage := "image"
+	ds.Spec.Template.Spec.Containers[0].Image = oldImage
+	ds, err := dsClient.Create(ds)
+	if err != nil {
+		t.Fatalf("failed to create daemonset %q: %v", dsName, err)
+	}
+
+	// Create 2 nodes
+	addNodes(nodeClient, 0, 2, nil, t)
+
+	// Validate everything works correctly, include marking pods
+	// as ready and ensure they launch on every node
+	validateDaemonSetPodsAndMarkReady(podClient, podInformer, 2, t)
+	validateDaemonSetStatus(dsClient, ds.Name, ds.Namespace, 2, t)
+
+	// Update the daemonset to trigger a rollout
+	newImage := "non-existent-image"
+	ds = updateDS(t, dsClient, dsName, func(update *v1beta1.DaemonSet) {
+		update.Spec.Template.Spec.Containers[0].Image = newImage
+	})
+
+	// Manually update image of first pod to the new image
+	// Note: daemonset image update does not propagate to its pods in integration test setting
```

Comment:
I didn't manage to find a way for daemonset image update to propagate to its pods without manually changing the pods. I even tried to delete the daemonset, but the daemonset failed to get deleted (you can still get the daemonset after the delete operation succeeds). This problem does not exist for deployment and replicaset, so I speculate that I miss out something important that is unique to daemonset.

As this test is about rolling update daemonset, the pods should be automatically deleted and replaced with new ones when there is a change to daemonset spec (correct me if i am wrong). However, coming out with something below does not work:
```
// the image change does not propagate to the pods
newDS := updateDS(t, dsClient, dsName, func(update *v1beta1.DaemonSet){
	update.Spec.Template.Spec.Containers[0].Image = newImage
})

// this checks whether there is at least one pod with new image,
// but the check fails (all pods still have old image)
checkAtLeastOnePodWithImage(podClient, newImage, t)
```

The pods are not deleted and replaced even after a long polling timeout (3 minutes). To get over it, I manually update the pod image to simulate the change, and the test passes (but this doesn't test the daemonset).

Reply:
@foxish Do you have any idea in solving problem mentioned above? Thanks!